src/main/java/com/ibm/darpc/DaRPCMemPool.java

```diff
@@ -0,0 +1,435 @@
+package com.ibm.darpc;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileChannel.MapMode;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+import com.ibm.disni.rdma.RdmaEndpoint;
+import com.ibm.disni.rdma.verbs.IbvMr;
+import com.ibm.disni.rdma.verbs.IbvPd;
+import com.ibm.disni.util.MemoryUtils;
+
+public class DaRPCMemPool {
+	final int defaultAllocationSize = 16 * 1024 * 1024; // 16MB
+	final int defaultMinAllocationSize = 4 * 1024; // 4KB
+	final int defaultAlignmentSize = 4 * 1024; // 4KB
+	final int defaultHugePageLimit = 0; // no huge pages by default
+
+	private HashMap<IbvPd, PdMemPool> pdMemPool; // One buddy allocator per protection domain
+	private LinkedList<IbvMr> mrs;
+	private boolean isOpen;
+
+	long currentRegion = 0;
+	private File dir;
+	protected int access; // RDMA access flags to use when registering memory regions
+	private long allocatedHugePageMemory;
+
+	// Configurable values
+	String hugePagePath = null;
+	protected int allocationSize;
+	protected int minAllocationSize;
+	protected int alignmentSize;
+	protected int hugePageLimit;
+
+
+
+
+
+	public DaRPCMemPool(String hugePagePath, int hugePageLimit, int allocationSize, int minAllocationSize, int alignmentSize) {
+		isOpen = false;
+		this.allocationSize = defaultAllocationSize;
+		this.minAllocationSize = defaultMinAllocationSize;
+		this.alignmentSize = defaultAlignmentSize;
+		this.hugePageLimit = defaultHugePageLimit;
+		this.access = IbvMr.IBV_ACCESS_LOCAL_WRITE | IbvMr.IBV_ACCESS_REMOTE_WRITE | IbvMr.IBV_ACCESS_REMOTE_READ;
+
+		this.hugePagePath = hugePagePath;
+
+		allocatedHugePageMemory = 0;
+
+		if (allocationSize >= 0) {
+			this.allocationSize = allocationSize;
+		}
+		if (minAllocationSize >= 0) {
+			this.minAllocationSize = minAllocationSize;
+		}
+		if (alignmentSize >= 0) {
+			this.alignmentSize = alignmentSize;
+		}
+		if (hugePageLimit >= 0) {
+			this.hugePageLimit = hugePageLimit;
+		}
+		if ((hugePagePath != null) && (!hugePagePath.equals("")) && (this.hugePageLimit > 0)) {
+			dir = new File(hugePagePath);
+			if (!dir.exists()){
+				dir.mkdirs();
+			}
+			for (File child : dir.listFiles()) {
+				child.delete();
+			}
+		} else {
+			this.hugePageLimit = 0;
```

Comment:
what happens if I pass a hugePageLimit but a null hugePagePath? Shouldn't we throw an error?